torchgeo/datamodules/vaihingen.py

```diff
@@ -45,117 +41,48 @@ def __init__(
         ``num_tiles_per_batch`` x ``num_patches_per_tile``.
 
         Args:
-            num_tiles_per_batch: The number of image tiles to sample from during
-                training
-            num_patches_per_tile: The number of patches to randomly sample from each
-                image tile during training
-            patch_size: The size of each patch, either ``size`` or ``(height, width)``.
-                Should be a multiple of 32 for most segmentation architectures
-            val_split_pct: The percentage of the dataset to use as a validation set
-            num_workers: The number of workers to use for parallel data loading
+            num_tiles_per_batch: Number of image tiles to sample from during training.
+            num_patches_per_tile: Number of patches to randomly sample from each image
+                tile during training.
+            patch_size: Size of each patch, either ``size`` or ``(height, width)``.
+                Should be a multiple of 32 for most segmentation architectures.
+            val_split_pct: Percentage of the dataset to use as a validation set.
+            num_workers: Number of workers for parallel data loading.
             **kwargs: Additional keyword arguments passed to
-                :class:`~torchgeo.datasets.Vaihingen2D`
+                :class:`~torchgeo.datasets.Vaihingen2D`.
 
         .. versionchanged:: 0.4
            *batch_size* was replaced by *num_tile_per_batch*, *num_patches_per_tile*,
            and *patch_size*.
         """
-        super().__init__()
+        super().__init__(Vaihingen2D, 1, num_workers, **kwargs)
 
-        self.num_tiles_per_batch = num_tiles_per_batch
+        self.train_batch_size = num_tiles_per_batch
         self.num_patches_per_tile = num_patches_per_tile
         self.patch_size = _to_tuple(patch_size)
         self.val_split_pct = val_split_pct
-        self.num_workers = num_workers
-        self.kwargs = kwargs
 
-        self.train_transform = AugmentationSequential(
-            Normalize(mean=0.0, std=255.0),
+        self.train_aug = AugmentationSequential(
+            K.Normalize(mean=self.mean, std=self.std),
             _RandomNCrop(self.patch_size, self.num_patches_per_tile),
             data_keys=["image", "mask"],
         )
-        self.test_transform = AugmentationSequential(
-            Normalize(mean=0.0, std=255.0),
+        self.aug = AugmentationSequential(
```

Comment:
The real Vaihingen2D dataset consists of images that are not of the same size. When lightning tries tries to collate `num_tiles_per_batch` scenes before the augmentations that make all inputs the same size are applied, the error occurs.

Reply:
Did we have this same problem for any other datamodules? Trying to remember if I've fixed this before...

Reply:
Oh, it was ChesapeakeCVPR where I had to do something like this. I'll try that approach.

Reply:
A few options I can use:

1. Remove `num_tiles_per_batch`, hard code `batch_size=1`
2. Apply `RandomNCrop` to sample instead of batch, dimensions may get tricky
3. Pad each sample to 3,816

Preferences?

Reply:
1. the random image sizes occur for all different data splits, so setting `batch_size=1` during training does not seem great. 
2. that could be a possibility
3. I don't think I like this approach, because in some cases the padding would be so extreme, also there might be other new datasets coming in that have a similar problem

Not sure what you think of this approach, but I tried the following where I define a collate_fn for the dataloader:
```
def custom_collate(batch):
    images = [item["image"] for item in batch]
    masks = [item["mask"] for item in batch]
    return {"image": images, "mask": masks}
```

And then in `on_after_batch_transfer`:
```
if self.trainer:
        if self.trainer.training:
            # Kornia requires masks to have a channel dimension
            batch["mask"] = [rearrange(item, "h w -> () h w") for item in batch["mask"]]
            augmented = []
            for img, mask in zip(batch["image"], batch["mask"]):
                augmented.append(self.train_transform({"image": img, "mask": mask}))
           batch = {
                "image": torch.cat([item["image"] for item in augmented], dim=0),
                "mask": torch.cat([item["mask"] for item in augmented], dim=0)
           }
```
I guess it defeats the purpose a bit of applying augmentations on GPU to a batch, but I thought it keeps the general new workflow of applying augmentations in torchgeo datamodules.

Reply:
1. I could do this for all stages, not just train

Reply:
Other benefit of 1 is that we can remove all this complicated "effective batch_size = num_tiles_per_batch x num_patches_per_tile" stuff and just have a single `batch_size` parameter that is the number of patches in a single tile. Each batch may be less diverse, but data loading should be faster, and the API becomes simpler.